microservices/payment-stripe/src/entities/coupon.ts

```diff
@@ -0,0 +1,116 @@
+import { IsTypeormDate, IsValidate } from '@lomray/microservice-helpers';
+import { IsEnum, IsInt, IsNumber, IsOptional, IsString, Length, Max, Min } from 'class-validator';
+import { JSONSchema } from 'class-validator-jsonschema';
+import {
+  Column,
+  CreateDateColumn,
+  Entity,
+  JoinTable,
+  ManyToMany,
+  OneToMany,
+  PrimaryColumn,
+  UpdateDateColumn,
+} from 'typeorm';
+import CouponDuration from '@constants/coupon-duration';
+import Product from '@entities/product';
+import PromoCode from '@entities/promo-code';
+
+@JSONSchema({
+  properties: {
+    products: { $ref: '#/definitions/Product', type: 'array' },
+    promoCodes: { $ref: '#/definitions/PromoCode', type: 'array' },
+  },
+})
+@Entity()
+class Coupon {
+  @JSONSchema({
+    description: 'Field for storing id of according coupon entity created on payment service side',
+  })
+  @PrimaryColumn({ type: 'varchar', length: 8 })
+  @Length(1, 8)
+  couponId: string;
+
+  @Column({ type: 'varchar', nullable: true, default: null })
+  @IsString()
+  @IsOptional()
+  userId: string | null;
+
+  @JSONSchema({
+    description: 'Name of the coupon displayed to customers on for instance invoices or receipts',
+  })
+  @Column({ type: 'varchar', nullable: true })
+  @IsString()
+  @IsOptional()
+  name: string | null;
+
+  @JSONSchema({
+    description: 'Amount that will be taken off the subtotal of any invoices for this customer.',
+  })
+  @Column({ type: 'float', nullable: true })
+  @IsNumber()
+  @IsOptional()
+  amountOff?: number;
+
+  @JSONSchema({
+    description:
+      'Percent that will be taken off the subtotal of any invoices for this customer for the duration of the coupon.',
+  })
+  @Column({ type: 'int', nullable: true })
+  @IsValidate(Coupon, (coupon) => !coupon.amountOff)
```

Comment:
<img width="682" alt="image" src="https://github.com/Lomray-Software/microservices/assets/44732463/04e41854-0e14-4f53-8411-3d9a9610352c">
decompose validation
```suggestion
  @IsValidate(Coupon, (coupon) => !Coupon.isAnountOff))
```